src/poetry/puzzle/provider.py

```diff
@@ -670,19 +670,12 @@ def complete_package(
 
             self.debug(f"<debug>Duplicate dependencies for {dep_name}</debug>")
 
-            non_direct_origin_deps: list[Dependency] = []
-            direct_origin_deps: list[Dependency] = []
-            for dep in deps:
-                if dep.is_direct_origin():
-                    direct_origin_deps.append(dep)
-                else:
-                    non_direct_origin_deps.append(dep)
-            deps = (
-                self._merge_dependencies_by_constraint(
-                    self._merge_dependencies_by_marker(non_direct_origin_deps)
+            dep_groups = self._group_by_source(deps)
```

Comment:
Similarly, a comment describing the intended merge logic (and why) would be valuable here. This refactoring is much more readable than the old version, and a comment will make it so even new contributors can digest it.